bolt-contracts/src/contracts/BoltValidatorsV2.sol

```diff
@@ -0,0 +1,383 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts/proxy/utils/UUPSUpgradeable.sol";
+
+import {BLS12381} from "../lib/bls/BLS12381.sol";
+import {BLSSignatureVerifier} from "../lib/bls/BLSSignatureVerifier.sol";
+import {IBoltValidatorsV2} from "../interfaces/IBoltValidatorsV2.sol";
+import {IBoltParametersV1} from "../interfaces/IBoltParametersV1.sol";
+
+/// @title Bolt Validators
+/// @notice This contract is responsible for registering validators and managing their configuration
+/// @dev This contract is upgradeable using the UUPSProxy pattern. Storage layout remains fixed across upgrades
+/// with the use of storage gaps.
+/// See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+/// To validate the storage layout, use the Openzeppelin Foundry Upgrades toolkit.
+/// You can also validate manually with forge: forge inspect <contract> storage-layout --pretty
+contract BoltValidatorsV2 is IBoltValidatorsV2, BLSSignatureVerifier, OwnableUpgradeable, UUPSUpgradeable {
+    using BLS12381 for BLS12381.G1Point;
+
+    // ========= STORAGE =========
+
+    /// @notice Bolt Parameters contract.
+    IBoltParametersV1 public parameters;
+
+    /// @notice Validators (aka Blockspace providers)
+    /// @dev Validators are blockspace providers for commitments.
+    ///
+    /// Validators in this mapping are identified by their sequence number.
+    /// The sequence number is an incremental index assigned to each validator
+    /// in the registry and is guaranteed to be unique.
+    mapping(uint32 => _Validator) internal VALIDATORS;
+
+    /// @notice Mapping of BLS public key hash to the sequence number of the validator 
+    /// in the VALIDATORS mapping. This mapping is used to quickly lookup a validator
+    /// by its BLS public key hash without iterating over the VALIDATORS mapping.
+    mapping(bytes20 => uint32) internal validatorPubkeyHashToSequenceNumber;
+
+    /// @notice Sequence number of the next validator to be registered
+    uint32 internal nextValidatorSequenceNumber;
+
+    /// @notice Mapping of controller index to its address. A controller map is 
+    /// used to identify entities with a 32-bit index instead of their full address.
+    mapping(uint32 => address) internal controllerIndexToAddress;
+    mapping(address => uint32) internal controllerAddressToIndex;
+    uint32 internal nextControllerIndex;
+
+    /// @notice Mapping of authorized operator index to its address. An authorized operator map is
+    /// used to identify entities with a 32-bit index instead of their full address.
+    mapping(uint32 => address) internal authorizedOperatorIndexToAddress;
+    mapping(address => uint32) internal authorizedOperatorToIndex;
+    uint32 internal nextAuthorizedOperatorIndex;
```

Comment:
this is bad, planning to clean this tomorrow

Reply:
Note that any new storage variables MUST be added to the end of the already existing list of storage variables btw